tempesta_fw/http_parser.c

```diff
@@ -4118,201 +4174,3191 @@ Req_Method_1CharStep: __attribute__((cold))
+				return CSTR_NEQ;
+
+			req->cond.flags |= TFW_HTTP_COND_ETAG_ANY;
+			__FSM_H2_I_MOVE_fixup(I_EoL, 1, 0);
+		}
+
+		if (IS_WS(c))
+			__FSM_H2_I_MOVE_NEQ_fixup(I_Etag, 1, 0);
+		return CSTR_NEQ;
+	}
+
+	__FSM_H2_TX_ETAG(I_Etag_W, '/', I_Etag_We);
+	__FSM_H2_TX_ETAG(I_Etag_We, '"', I_Etag_Weak);
+
+	__FSM_STATE(I_Etag_Weak) {
+		parser->hdr.flags |= TFW_STR_ETAG_WEAK;
+		__FSM_JMP(I_Etag_Val);
+	}
+
+	__FSM_STATE(I_Etag_Val) {
+		weak = parser->hdr.flags & TFW_STR_ETAG_WEAK;
+		__FSM_H2_I_MATCH_MOVE_NEQ_fixup(token, I_Etag_Val,
+						(TFW_STR_VALUE | weak));
+		c = *(p + __fsm_sz);
+		if (likely(c == '"')) {
+			parser->hdr.flags &= ~TFW_STR_ETAG_WEAK;
+			__FSM_H2_I_MOVE_fixup(I_EoT, __fsm_sz + 1,
+					      TFW_STR_VALUE | weak);
+		}
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(I_EoT) {
+		if (IS_WS(c))
+			__FSM_H2_I_MOVE_fixup(I_EoT, 1, 0);
+		if (c == ',')
+			__FSM_H2_I_MOVE_NEQ_fixup(I_Etag, 1, 0);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(I_EoL) {
+		if (IS_WS(c))
+			__FSM_H2_I_MOVE_fixup(I_EoL, 1, 0);
+		return CSTR_NEQ;
+	}
+
+done:
+	return r;
+}
+STACK_FRAME_NON_STANDARD(__h2_req_parse_if_none_match);
+
+static int
+__h2_req_parse_host(TfwHttpReq *req, unsigned char *data, size_t len, bool fin)
+{
+	int r = CSTR_NEQ;
+	__FSM_DECLARE_VARS(req);
+
+	__FSM_START(parser->_i_st);
+
+	__FSM_STATE(Req_I_H_Start) {
+		if (likely(isalnum(c) || c == '.' || c == '-'))
+			__FSM_H2_I_MOVE(Req_I_H);
+		if (likely(c == '['))
+			__FSM_H2_I_MOVE_NEQ(Req_I_H_v6, 1);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_H) {
+		/* See Req_AuthorityGen processing. */
+		if (likely(isalnum(c) || c == '.' || c == '-'))
+			__FSM_H2_I_MOVE(Req_I_H);
+		if (c == ':')
+			__FSM_H2_I_MOVE(Req_I_H_Port);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_H_v6) {
+		/* See Req_AuthorityIPv6 processing. */
+		if (likely(isxdigit(c) || c == ':'))
+			__FSM_H2_I_MOVE_NEQ(Req_I_H_v6, 1);
+		if (likely(c == ']'))
+			__FSM_H2_I_MOVE(Req_I_H_v6_End);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_H_v6_End) {
+		if (likely(c == ':'))
+			__FSM_H2_I_MOVE(Req_I_H_Port);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_H_Port) {
+		/* See Req_Port processing. */
+		if (likely(isdigit(c)))
+			__FSM_H2_I_MOVE(Req_I_H_Port);
+		return CSTR_NEQ;
+	}
+
+done:
+	return r;
+}
+STACK_FRAME_NON_STANDARD(__h2_req_parse_host);
+
+static int
+__h2_req_parse_referer(TfwHttpMsg *hm, unsigned char *data, size_t len,
+		       bool fin)
+{
+	int r = CSTR_NEQ;
+	__FSM_DECLARE_VARS(hm);
+
+	__FSM_START(parser->_i_st);
+
+	__FSM_STATE(Req_I_Referer) {
+		__FSM_H2_I_MATCH_MOVE(uri, Req_I_Referer);
+		if (IS_WS(*(p + __fsm_sz)))
+			__FSM_H2_I_MOVE_n(Req_I_EoT, __fsm_sz + 1);
+		return CSTR_NEQ;
+	}
+	__FSM_STATE(Req_I_EoT) {
+		if (IS_WS(c))
+			__FSM_H2_I_MOVE(Req_I_EoT);
+		return CSTR_NEQ;
+	}
+
+done:
+	return r;
+}
+STACK_FRAME_NON_STANDARD(__h2_req_parse_referer);
+
+static int
+__h2_parse_http_date(TfwHttpMsg *hm, unsigned char *data, size_t len, bool fin)
+{
```

Comment:
>It seems we have only 3 options:

Yeah - in my first attempt of HTTP/2-parser implementation I tried to use approach from p.3, but was stumbled upon two problems:
1. A lot of additional branching and non-optimal code for both: HTTP/2 and HTTP/1.1 versions.
2. More complexity had been added into the already rather complicated  parser, and the code became almost unreadable; it became too hard to understand what is going on in the parser at all, and it would be very difficult to maintain and extend in the future (it seems to me, that the variant from p.2 will have the same problem).

Thus, I decided to rework the HTTP/2-parser to make it completely separated.

Reply:
Done in b1e6a1149c2d106234644e035a18906a3ae15cf1